### DISCUSSION TITLE
How to build userland tools seperately

### DISCUSSION THREAD

**Discussion by furbyhaxx** [openzfs/zfs#15794] [Q&A] [answered]
I'm a bit lost, I compiled zfs into the kernel with

```
./configure --enable-linux-builtin
..
./copy-builtin /usr/src/linux
```

added required config files and compiled kernel, all good so far.

But how do I get the userland tools compiled? If I just do:
```
./configure --enable-linux-builtin
make
make install
```

the compiled tools complain about the kernel module not being loaded.

**Comment by rincebrain** [comment]
How you compile them, at the moment, has no bearing on how they decide the module is loaded, so if it's complaining the module isn't loaded, it's probably because it's not loaded. Does /proc/spl/kstat/zfs exist? If not, I would suggest your kernel is missing ZFS.

**Reply by furbyhaxx** [reply]
you're right, that does not exist. But I don't get it why it isn't, copy-builtin, add CONFIG_ZFS=y and make. am I missing something here? I do not get any errors.

**Reply by rincebrain** [reply, marked as answer]
Commonly, because CONFIG_ZFS=y got erased at some point, and so it didn't actually get built in. If you have a copy of your running kernel's configure, from /proc/config.gz or /boot/config-something or w/e, you might want to check what it actually produced.

**Reply by furbyhaxx** [reply]
I completely rebuilt my kernel and config and now it works. There was a config incompatibility which always reset the config to defaults.
Thanks for pointing in the right direction.

**Comment by IvanVolosyuk** [comment]
There was a bug in install scripts for some versions of ZFS with new
kernels. Check that fs/Makefile has a reference to ZFS directory.

**Reply by furbyhaxx** [reply]
yes it has a reference to the zfs dir
